infrastructure/equinix-metal/main.tf

```diff
@@ -0,0 +1,157 @@
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "curl -sfL https://get.k3s.io | INSTALL_K3S_CHANNEL=${var.k3s_version} K3S_TOKEN=${var.k3s_token} sh -s - server --node-taint node-role.kubernetes.io/control-plane=:NoSchedule --flannel-backend=none --disable-network-policy --disable=traefik",
+      "systemctl is-active --quiet k3s.service",
+    ]
+  }
+}
+
+// NOTE: to extract KUBECONFIG we need to copy kubeconfig in controlplane:/etc/rancher/k3s/k3s.yaml
+
+resource "equinix_metal_device" "worker" {
+  for_each            = toset(var.worker_nodes)
+  hostname            = "${var.cluster_name}-${each.value}"
+  plan                = var.device_plan
+  metro               = var.device_metro
+  operating_system    = var.device_os
+  billing_cycle       = var.billing_cycle
+  project_id          = var.equinix_project_id
+  project_ssh_key_ids = [equinix_metal_project_ssh_key.ssh_key.id]
+  depends_on          = [equinix_metal_device.control_plane]
+  user_data           = <<EOF
+#!/bin/bash
+curl -sfL https://get.k3s.io | INSTALL_K3S_CHANNEL="${var.k3s_version}" sh -s - agent --token "${var.k3s_token}" --server "https://${equinix_metal_device.control_plane.access_private_ipv4}:6443"
+EOF
+
+  behavior {
+    allow_changes = [
+      "custom_data",
+      "user_data"
+    ]
+  }
+}
+
+resource "null_resource" "install_cilium_cni" {
+  depends_on          = [equinix_metal_device.control_plane]
+  triggers = {
+    always_run = "${timestamp()}"
+  }
+
+  connection {
+    user = "root"
+    private_key = file(var.ssh_private_key_path)
+    host = equinix_metal_device.control_plane.access_public_ipv4
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "echo '@@@@@@ Installing Cilium @@@@@@'",
+      "CILIUM_CLI_VERSION=$(curl -s https://raw.githubusercontent.com/cilium/cilium-cli/main/stable.txt)",
+      "CLI_ARCH=amd64",
+      "curl -L --fail --remote-name-all https://github.com/cilium/cilium-cli/releases/download/$${CILIUM_CLI_VERSION}/cilium-linux-$${CLI_ARCH}.tar.gz{,.sha256sum}",
+      "sha256sum --check cilium-linux-$${CLI_ARCH}.tar.gz.sha256sum",
+      "sudo tar xzvfC cilium-linux-$${CLI_ARCH}.tar.gz /usr/local/bin",
+      "rm cilium-linux-$${CLI_ARCH}.tar.gz{,.sha256sum}",
+      "echo '@@@@@@ Installed Cilium @@@@@@'",
+      "export KUBECONFIG=/etc/rancher/k3s/k3s.yaml",
+      "echo '@@@@@@ Adding Cilium CNI to cluster @@@@@@'",
+      "cilium install --version ${var.cilium_version}",
+      "cilium status --wait"
+    ]
+  }
+}
+
+resource "null_resource" "flux_env_vars" {
+  depends_on          = [null_resource.install_cilium_cni]
+  triggers = {
+    always_run = "${timestamp()}"
+  }
+
+  connection {
+    user = "root"
+    private_key = file(var.ssh_private_key_path)
+    host = equinix_metal_device.control_plane.access_public_ipv4
+  }
+
+  provisioner "file" {
+    content = <<EOF
+GITHUB_TOKEN=${var.flux_github_token}
+KUBECONFIG=/etc/rancher/k3s/k3s.yaml
+EOF
+    destination = "/tmp/flux_env_vars"
+  }
+}
```

Comment:
@rossf7 if we are copying the kubeconfig from the controlplane then we can rename resource block to `get-kubeconfig`?

also should we use this kubeconfig to install kepler and prometheus-stack with grafana dashboard or will plan it in some other workflow?

Reply:
@dipankardas011 The separate block is mainly for the github token. This allows us to view logs from the `flux bootstrap` command. Otherwise tofu hides them to avoid outputting the token.

For installing kepler and prometheus as far as possible we should apply all changes via flux. Flux is granted `cluster-admin` privileges as part of the bootstrap process. https://github.com/cncf-tags/green-reviews-tooling/blob/86f136fb88e77a6743dcec6acc4ca83e1d484f7c/clusters/flux-system/gotk-components.yaml#L247

We still need the kubeconfig for manual trouble shooting but anyone with ssh access can get it and generate a readonly kubeconfig using the steps you're working on.

Reply:
got it thanks @rossf7  for clarifying :)